Implement pagination for GET /users endpoint #41

amirabramovich opened the pull request:
🚀 Description

added ?limit=X and ?start=X path variables for GET /users API.
also refactored a bit to make the code prettier :)

## 📄 Motivation and Context

solving issue [#41 ](https://github.com/jmprathab/MyHome/issues/41)

amirabramovich committed:
Implement pagination for GET /users endpoint #41

jmprathab commented on src/main/java/com/myhome/services/springdatajpa/UserSDJpaService.java:
```diff
@@ -39,33 +40,31 @@
-    return userListSet;
+    Stream<User> userStream = userListSet.stream();
+    if (start != null) {
+      userStream = userStream.filter(user -> user.getUserId().startsWith(String.valueOf(start)));
```
Implementation of `limit` seems to be fine. But, start ideally has to be used for pagination. Let me try to explain it with a use case.

User hits `/users?limit=20&start=0` - Display the first 20 users.
User hits `/users?limit=20&start=20` - Return the next 20 users.

Considering your current implementation, it seems ideal to rename the `start` parameter to `filter`.

amirabramovich committed:
fixed start definition

amirabramovich committed:
implemented basic class for controller tests and added JUnit tests for User Controller

amirabramovich commented:
I fixed the 'start', i misunderstood the definition of it in the begining, now it is ok.
Also, I added basic test utils for controllers and specific tests for User controller, create is pass fast and good, but GET APIs returns 403 for some reason, im not sure if it is bug in my tests or there is a bug in production code.

jmprathab commented on src/test/java/com/myhome/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,66 @@
+
+  @Test
+  @DisplayName("test get /users listAllUsers() positive")
+  void listAllUsersPositive() {
```
I am not wrong, except for `POST to /users/login` `POST to /users` and `GET to /status` all other endpoints require JWT to be passed in request header.

A JWT can be obtained with `/users/login`. Look for `token` header in the HTTP response, that is your JWT.

amirabramovich committed:
fixed JUnits with jwt-token

amirabramovich commented:
i fixed tests which required jwt-tokens, now everything pass :)
i recommand to open an issue per controller for adding controller tests and describe testing flows.
thanks for ControllerTestBase class which i created, testing flow for all controllers should be pretty easy and fast.
it is fine by me to merge this PR and close the original issue, since it is solved.

jmprathab reviewed (commented):
Looks good

jmprathab commented:
@amirabramovich We do have an issue for adding test cases. jmprathab/MyHome#38

jmprathab commented:
@amirabramovich I had to revert the pull request from master branch. It seems like `gradlew build` is failing due to missing dependencies. Can you please take a look?

Can you please raise another PR?
